File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */

var maxSubarrayLength = function(nums,k) {
    let n=nums.length;
    let freq=new Map();

    let i=0;
    let j=0;
    let res=0;

    while(j<n){
        freq.set(nums[j],(freq.get(nums[j])||0)+1);

        while(freq.get(nums[j])>k){

            freq.set(nums[i],freq.get(nums[i])-1);
            i++;
        }
        res=Math.max(res,j-i+1);
        j++;
    }
    return res;
};

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.go
func maxSubarrayLength(nums []int,k int) int {
    n:=len(nums)
    freq:=make(map[int]int)
    i:=0
    j:=0
    res:=0

    for j<n {
        freq[nums[j]]++

        for freq[nums[j]]>k {
            freq[nums[i]]--
            i++
        }

        if j-i+1>res {
            res=j-i+1
        }

        j++
    }
    return res
}